test/lit/passes/precompute-gc.wast

```diff
@@ -34,4 +38,267 @@
+  ;; elsewhere (we'd need immutability or escape analysis)
+  (call $log
+   (struct.get $struct 0 (local.get $x))
+  )
+  ;; Assign a new struct
+  (local.set $x
+   (struct.new_with_rtt $struct
+    (i32.const 2)
+    (rtt.canon $struct)
+   )
+  )
+  (call $log
+   (struct.get $struct 0 (local.get $x))
+  )
+  ;; Assign a new value
+  (struct.set $struct 0
+   (local.get $x)
+   (i32.const 3)
+  )
+  (call $log
+   (struct.get $struct 0 (local.get $x))
+  )
+ )
+ ;; CHECK:      (func $load-from-struct-bad-merge (param $i i32)
+ ;; CHECK-NEXT:  (local $x (ref null $struct))
+ ;; CHECK-NEXT:  (if
+ ;; CHECK-NEXT:   (local.get $i)
+ ;; CHECK-NEXT:   (local.set $x
+ ;; CHECK-NEXT:    (struct.new_with_rtt $struct
+ ;; CHECK-NEXT:     (i32.const 1)
+ ;; CHECK-NEXT:     (rtt.canon $struct)
+ ;; CHECK-NEXT:    )
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:   (local.set $x
+ ;; CHECK-NEXT:    (struct.new_with_rtt $struct
+ ;; CHECK-NEXT:     (i32.const 2)
+ ;; CHECK-NEXT:     (rtt.canon $struct)
+ ;; CHECK-NEXT:    )
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (call $log
+ ;; CHECK-NEXT:   (struct.get $struct 0
+ ;; CHECK-NEXT:    (local.get $x)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $load-from-struct-bad-merge (param $i i32)
+  (local $x (ref null $struct))
+  ;; a merge of two different $x values cannot be precomputed
+  (if
+   (local.get $i)
+   (local.set $x
+    (struct.new_with_rtt $struct
+     (i32.const 1)
+     (rtt.canon $struct)
+    )
+   )
+   (local.set $x
+    (struct.new_with_rtt $struct
+     (i32.const 2)
+     (rtt.canon $struct)
+    )
+   )
+  )
+  (call $log
+   (struct.get $struct 0 (local.get $x))
+  )
+ )
+ ;; CHECK:      (func $modify-gc-heap (param $x (ref null $struct))
+ ;; CHECK-NEXT:  (struct.set $struct 0
+ ;; CHECK-NEXT:   (local.get $x)
+ ;; CHECK-NEXT:   (i32.add
+ ;; CHECK-NEXT:    (struct.get $struct 0
+ ;; CHECK-NEXT:     (local.get $x)
+ ;; CHECK-NEXT:    )
+ ;; CHECK-NEXT:    (i32.const 1)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $modify-gc-heap (param $x (ref null $struct))
+  (struct.set $struct 0
+   (local.get $x)
+   (i32.add
+    (struct.get $struct 0
+     (local.get $x)
+    )
+    (i32.const 1)
+   )
+  )
+ )
+ ;; --fuzz-exec verifies the output of this function, checking that the change
+ ;; makde in modify-gc-heap is not ignored
+ (func "load-from-struct-bad-escape"
+  (local $x (ref null $struct))
+  (local.set $x
+   (struct.new_with_rtt $struct
+    (i32.const 1)
+    (rtt.canon $struct)
+   )
+  )
+  (call $modify-gc-heap
+   (local.get $x)
+  )
+  (call $log
+   (struct.get $struct 0 (local.get $x))
+  )
+ )
+ ;; CHECK:      (func $load-from-struct-bad-arrive (param $x (ref null $struct))
+ ;; CHECK-NEXT:  (call $log
+ ;; CHECK-NEXT:   (struct.get $struct 0
+ ;; CHECK-NEXT:    (local.get $x)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $load-from-struct-bad-arrive (param $x (ref null $struct))
```

Comment:
Is this just testing that no precomputation happens when the struct is a parameter?

Reply:
yes